### DISCUSSION TITLE
Exercice 2: TDD et Refacto, pour du Clean Code

### DISCUSSION THREAD

**Discussion by guillaumeagile** [guillaumeagile/LPRO3_DQL_2021#16] [Show and tell]
Voici l'exercice:

## opérations GIT
allez voir le code ici:
https://github.com/guillaumeagile/LPRO3_DQL_2021/tree/main/EXERCICE_2

Vous pouvez garder le répertoire local que vous avez créé pour l'exercice précédent, mais il vous faut revenir sur la branche MAIN et la mettre à jour sur votre machine
> git checkout main
git pull

De là (et uniquement en ayant fait les opérations précédentes), vous avez 2 possibilités: 
- créer une nouvelle branche de travail depuis main
> git checkout -b EXO2_VotreNom

- ou bien repartir de votre branche à vous que vous aviez préparer mais il faut la rebase avec master (remonter les dernières modifications qui sont dans main vers votre ancienne branche)
> git checkout NOM_DE_VOTRE_ANCIENNE_BRANCHE
git rebase main

Si il y a des conflits, faire
> git rebase --continue

Pour mettre votre branche à jour, il faut finir par
> git push -f

Si vous galérez trop, repartez de zéro, dans un nouveau dossier vide et faites
> git clone https://github.com/guillaumeagile/LPRO3_DQL_2021.git

## ouvrir le projet dans un IDE

Ouvrez le répertoire  LPRO3_DQL_2021/EXERCICE_2/  avec [IntelliJ IDEA Community Edition](https://www.jetbrains.com/idea/download/#section=windows)

Utilisez Java 17 LTS   ou Java 15
https://www.oracle.com/java/technologies/downloads/

# EXERCICE
il y a dans src/test/java/LPRO_DQL/ déjà 2 tests

Le code est dans la fonction f1,  il fonctionne mais il doit être refactoré.

Pour vous assurer de ne rien casser pendant le refactoring, vous devez écrire suffisamment de tests pour couvrir les cas "intéressants".
C'est à dire les cas qui révèlent un comportement différent de la fonction selon les inputs que l'on lui passe.
**Vous devez donc écrire "les bons tests"**, ceux qui reflètent une situation à chaque fois différente (pas la peine de tester deux fois la même chose ou des cas similaires).
https://medium.com/@Cyrdup/unit-testing-youre-doing-it-wrong-407a07692989

Etant donné que le code fonctionne, pour écrire un nouveau test, copiez coller le test précédent, changez quelque chose dans les paramètres d'appel de la fonction f1
Observez le résultat "actual",  le test va "fail" si vous avez effectivement introduit une nouvelle situation en changeant le  ou les paramètres.
SI le comportement de votre nouveau test vous semble cohérent, utilisez la valeur "actual" pour changer "expected" afin que actual et expected soient égaux.
Ca vous donne un test vert.
Attention, vous ne pouvez faire comme ca que parce que vous savez que la fonction f1 fonctionne bien.
Continuez comme ca jusqu'à ce vous pensez que vous avez fait le tour de toutes les situations intéressantes à tester.

_Le Golden Master Testing permet de sécuriser la refonte d’une application en appliquant une couche de test sur l’application initiale avant toute modification._

IMPORTANT : à chaque fois que vous écrivez un test vert, faites un commit !

Une fois les bons tests en place (et verts), vous pouvez vous relancer dans le refactoring, c'est à dire chasser les code smells
En commençant par renommer les variables.
Mais ce n'est pas tout, ensuite il vous faut essayer de diminuer la complexité et si possible  le nombre de ligne de code.
Des exemples de code smell:
https://refactoring.guru/refactoring/smells
https://maximilianocontieri.com/series/code-smells

A vous de jouer

Conclusion:
![image](https://user-images.githubusercontent.com/197786/142411188-26f3d043-5e43-43f5-85cd-85b002174471.png)

**Comment by TBoltsp** [comment]
Bonjour, voici le lien d'accès au serveur discord (rudimentaire) de la promo LP_DQL

https://discord.gg/P3QjUJMRC9

**Reply by guillaumeagile** [reply]
Merci

**Comment by guillaumeagile** [comment]
Pour vous aider à préparer l'exercice de refactoring, 
jetez un œil à l'utilisation des Streams  pour faire des filtres sur des listes en Java
https://www.baeldung.com/java-filter-collection-by-list
et aussi
https://www.baeldung.com/java-stream-filter-lambda

Ca vous explique aussi comment utiliser les Lambda ...

**Reply by guillaumeagile** [reply]
Et un tuto sur les HashMap (que j'utilise dans le code et les tests)
https://www.data-transitionnumerique.com/java-hashmap/

**Comment by bas22tien** [comment]
Bonjour, monsieur auriez vous un tuto pour bien configurer IntelliJ dans le but de faire l'exercice.

**Reply by guillaumeagile** [reply]
voici
https://blog.jetbrains.com/idea/2020/09/java-15-and-intellij-idea/#:~:text=IntelliJ%20IDEA%20Configuration&text=You%20can%20configure%20it%20to,your%20Project%20and%20Modules%20settings.
